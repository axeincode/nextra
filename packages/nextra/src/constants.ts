import path from 'node:path'

export const MARKDOWN_EXTENSION_REGEX = /\.mdx?$/

export const IS_PRODUCTION = process.env.NODE_ENV === 'production'

export const LOCALE_REGEX = /\.([a-z]{2}(-[A-Z]{2})?)$/

export const DEFAULT_LOCALE = 'en-US'

export const DEFAULT_CONFIG = {
  staticImage: true,
  flexsearch: {
    codeblocks: true
  }
}

export const OFFICIAL_THEMES = [
  'nextra-theme-docs',
  'nextra-theme-blog'
] as const

export const META_FILENAME = '_meta.json'

export const CWD = process.cwd()

export const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const

export const PUBLIC_DIR = path.join(CWD, 'public')

export const CACHE_DIR = path.join(CWD, '.next', 'cache')

export const ASSET_DIR = path.join(CWD, '.next', 'static', 'chunks')

export const EXTERNAL_URL_REGEX = /^https?:\/\//
