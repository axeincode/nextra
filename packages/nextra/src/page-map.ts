import { PageMapItem } from './types'
import { parseFileName, existsSync } from './utils'
import path from 'path'
import filterRouteLocale from './filter-route-locale'

export function findPagesDir(dir = process.cwd()): string {
  // prioritize ./pages over ./src/pages
  if (existsSync(path.join(dir, 'pages'))) return 'pages'
  if (existsSync(path.join(dir, 'src/pages'))) return 'src/pages'

  throw new Error(
    "> Couldn't find a `pages` directory. Please create one under the project root"
  )
}

export function getPageMap(
  currentResourcePath: string,
  pageMap: PageMapItem[],
  fileMap: Record<string, PageMapItem>,
  defaultLocale: string
): [PageMapItem[], string, string] {
  const activeRouteLocale = parseFileName(currentResourcePath).locale
  const pageItem = fileMap[currentResourcePath]
  const metaPath = path.dirname(currentResourcePath)
  const metaExtension = activeRouteLocale ? `${activeRouteLocale}.json` : `json`
  const pageMeta =
    fileMap[`${metaPath}/meta.${metaExtension}`]?.meta?.[pageItem.name]
  const title =
    (typeof pageMeta === 'string' ? pageMeta : pageMeta?.title) || pageItem.name

  return [
    activeRouteLocale
      ? filterRouteLocale(pageMap, activeRouteLocale, defaultLocale)
      : pageMap,
    pageItem.route,
    title
  ]
}
