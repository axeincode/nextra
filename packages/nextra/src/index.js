import { NextraPlugin, pageMapCache } from './plugin'
import { DEFAULT_LOCALE, MARKDOWN_EXTENSION_REGEX } from './constants'

const DEFAULT_EXTENSIONS = ['js', 'jsx', 'ts', 'tsx']
const MARKDOWN_EXTENSIONS = ['md', 'mdx']

const nextra = (...config) =>
  function withNextra(nextConfig = {}) {
    const nextraConfig =
      typeof config[0] === "string"
        ? {
            theme: config[0],
            themeConfig: config[1]
          }
        : config[0]

    const nextraPlugin = new NextraPlugin(nextraConfig)
    const { pageExtensions = DEFAULT_EXTENSIONS } = nextConfig

    if (nextConfig.i18n?.locales) {
      console.log(
        '[nextra] You have Next.js i18n enabled, read here https://nextjs.org/docs/advanced-features/i18n-routing for the docs.'
      )
    }

    // If `i18n.locales` and `i18n.defaultLocale` were not specified,
    // client will receive error - Text content does not match server-rendered HTML.
    // Due to `const { locale } = useRouter()` where `locale` will be `undefined`
    // To fix it we need to explicitly specify `i18n.locales` and `i18n.defaultLocale`
    const i18n = {
      locales: nextConfig.i18n?.locales || [DEFAULT_LOCALE],
      defaultLocale: nextConfig.i18n?.defaultLocale || DEFAULT_LOCALE
    }

    return {
      ...nextConfig,
      i18n,
      pageExtensions: [...pageExtensions, ...MARKDOWN_EXTENSIONS],
      webpack(config, options) {
        config.plugins ||= []
        config.plugins.push(nextraPlugin)

        config.module.rules.push({
          test: MARKDOWN_EXTENSION_REGEX,
          use: [
            options.defaultLoaders.babel,
            {
              loader: 'nextra/loader',
              options: {
                ...nextraConfig,
                locales: i18n.locales,
                defaultLocale: i18n.defaultLocale,
                pageMapCache
              }
            }
          ]
        })

        if (typeof nextConfig.webpack === 'function') {
          return nextConfig.webpack(config, options)
        }

        return config
      }
    }
  }

module.exports = nextra
