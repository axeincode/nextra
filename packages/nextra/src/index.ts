import { NextraPlugin, pageMapCache } from './plugin'
import { DEFAULT_LOCALE, MARKDOWN_EXTENSION_REGEX } from './constants'
import { LoaderOptions, Nextra } from './types'

const DEFAULT_EXTENSIONS = ['js', 'jsx', 'ts', 'tsx'] as const
const MARKDOWN_EXTENSIONS = ['md', 'mdx'] as const

const nextra: Nextra = (...args) =>
  function withNextra(nextConfig = {}) {
    const nextraConfig =
      typeof args[0] === 'string'
        ? {
            theme: args[0],
            themeConfig: args[1] as string
          }
        : args[0]

    const nextraPlugin = new NextraPlugin(nextraConfig)
    const { i18n, pageExtensions = DEFAULT_EXTENSIONS } = nextConfig

    if (i18n?.locales) {
      console.log(
        '[nextra] You have Next.js i18n enabled, read here https://nextjs.org/docs/advanced-features/i18n-routing for the docs.'
      )
    } else if (!i18n?.defaultLocale) {
      // If `i18n.locales` and `i18n.defaultLocale` were not specified,
      // client will receive error - Text content does not match server-rendered HTML.
      // Due to `const { locale } = useRouter()` where `locale` will be `undefined`
      // To fix it we need to explicitly specify `i18n.locales` and `i18n.defaultLocale`
      nextConfig.i18n = {
        ...i18n,
        locales: [DEFAULT_LOCALE],
        defaultLocale: DEFAULT_LOCALE
      }
    }

    return {
      ...nextConfig,
      pageExtensions: [...pageExtensions, ...MARKDOWN_EXTENSIONS],
      webpack(config, options) {
        config.plugins ||= []
        config.plugins.push(nextraPlugin)

        config.module.rules.push({
          test: MARKDOWN_EXTENSION_REGEX,
          use: [
            options.defaultLoaders.babel,
            {
              loader: 'nextra/loader',
              options: <LoaderOptions>{
                ...nextraConfig,
                locales: nextConfig.i18n?.locales,
                defaultLocale: nextConfig.i18n?.defaultLocale,
                pageMapCache
              }
            }
          ]
        })

        if (typeof nextConfig.webpack === 'function') {
          return nextConfig.webpack(config, options)
        }

        return config
      }
    }
  }

module.exports = nextra
