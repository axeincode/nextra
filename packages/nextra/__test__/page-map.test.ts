import { describe, it, expect, beforeAll } from 'vitest'
import path from 'node:path'
import { resolvePageMap } from '../src/page-map'
import { collectFiles } from '../src/plugin'
import { FileMap, PageMapItem } from '../src/types'
import { CWD } from '../src/constants'

const PAGES_DIR = path.join(CWD, '..', '..', 'examples', 'swr-site', 'pages')
const filePath = (name: string) => path.join(PAGES_DIR, name)

const defaultLocale = 'en-US'

describe('Page Process', () => {
  let pageMap: PageMapItem[]
  let fileMap: FileMap
  beforeAll(async () => {
    const { items, fileMap: data } = await collectFiles(PAGES_DIR, [''])
    pageMap = items
    fileMap = data
  })

  it('pageMap en-US', () => {
    const indexData = resolvePageMap({
      filePath: filePath('docs/data-fetching.en-US.mdx'),
      items: pageMap,
      fileMap,
      defaultLocale
    })
    expect([indexData.pageMap, indexData.route]).toMatchSnapshot()

    const gettingStartData = resolvePageMap({
      filePath: filePath('docs/getting-started.en-US.mdx'),
      items: pageMap,
      fileMap,
      defaultLocale
    })
    expect(gettingStartData.pageMap).toEqual(indexData.pageMap)
  })

  it('pageMap ru', () => {
    const indexData = resolvePageMap({
      filePath: filePath('docs/data-fetching.ru.mdx'),
      items: pageMap,
      fileMap,
      defaultLocale
    })
    expect([indexData.pageMap, indexData.route]).toMatchSnapshot()

    const gettingStartData = resolvePageMap({
      filePath: filePath('docs/getting-started.ru.mdx'),
      items: pageMap,
      fileMap,
      defaultLocale
    })
    expect(gettingStartData.pageMap).toEqual(indexData.pageMap)
  })
})
