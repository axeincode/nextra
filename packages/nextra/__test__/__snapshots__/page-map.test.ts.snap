// Vitest Snapshot v1

exports[`Page Process > pageMap en-US 1`] = `
[
  [
    {
      "kind": "MdxPage",
      "name": "404",
      "route": "/404",
    },
    {
      "kind": "MdxPage",
      "name": "500",
      "route": "/500",
    },
    {
      "data": {
        "404": {
          "theme": {
            "timestamp": false,
            "typesetting": "article",
          },
          "type": "page",
        },
        "500": {
          "theme": {
            "timestamp": false,
            "typesetting": "article",
          },
          "type": "page",
        },
        "about": {
          "items": {
            "a-page": {
              "title": "A Page",
            },
            "acknowledgement": {
              "title": "Acknowledgement",
            },
            "contributors": {
              "href": "https://github.com/vercel/swr/graphs/contributors",
              "newWindow": true,
              "title": "Contributors",
            },
            "team": {
              "title": "Team",
            },
          },
          "title": "About",
          "type": "menu",
        },
        "blog": {
          "theme": {
            "breadcrumb": false,
            "footer": true,
            "sidebar": false,
            "toc": true,
            "typesetting": "article",
          },
          "title": "Blog",
          "type": "page",
        },
        "docs": {
          "theme": {
            "toc": true,
          },
          "title": "Docs",
          "type": "page",
        },
        "examples": {
          "theme": {
            "layout": "full",
          },
          "title": "Examples",
          "type": "page",
        },
        "index": {
          "display": "hidden",
          "title": "Introduction",
          "type": "page",
        },
        "nextra_link": {
          "href": "https://github.com/shuding/nextra",
          "newWindow": true,
          "title": "Nextra ‚Üó",
          "type": "page",
        },
      },
      "kind": "Meta",
      "locale": "en-US",
    },
    {
      "children": [
        {
          "data": {
            "a-page": {
              "display": "hidden",
              "theme": {
                "layout": "raw",
              },
              "title": "A Page",
              "type": "page",
            },
            "acknowledgement": "üß© Acknowledgement",
            "team": "üë• Team",
          },
          "kind": "Meta",
          "locale": "en-US",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "a-page",
          "route": "/about/a-page",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "acknowledgement",
          "route": "/about/acknowledgement",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "team",
          "route": "/about/team",
        },
      ],
      "kind": "Folder",
      "name": "about",
      "route": "/about",
    },
    {
      "children": [
        {
          "data": {
            "swr-v1": {
              "theme": {
                "breadcrumb": true,
              },
              "title": "Announcing SWR 1.0",
            },
          },
          "kind": "Meta",
          "locale": "en-US",
        },
        {
          "frontMatter": {
            "description": "Almost 2 years ago we open sourced SWR, the tiny data-fetching React library that people love. Today we are reaching another milestone: the 1.0 version of SWR.",
            "image": "https://assets.vercel.com/image/upload/v1630059453/swr/v1.png",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "swr-v1",
          "route": "/blog/swr-v1",
        },
      ],
      "kind": "Folder",
      "name": "blog",
      "route": "/blog",
    },
    {
      "frontMatter": {
        "searchable": false,
      },
      "kind": "MdxPage",
      "locale": "en-US",
      "name": "blog",
      "route": "/blog",
    },
    {
      "children": [
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "404-500",
          "route": "/docs/404-500",
        },
        {
          "data": {
            "--- hey": {
              "title": "Getting Started",
              "type": "separator",
            },
            "--- my_new_separator": {
              "title": "Advanced",
              "type": "separator",
            },
            "404-500": "404/500 Custom Error Pages",
            "advanced": {
              "title": "Advanced",
            },
            "arguments": "Arguments",
            "callout": "Callout",
            "change-log": {
              "theme": {
                "sidebar": false,
              },
              "title": "Change Log",
            },
            "code-block-without-language": "Code Block without Language",
            "conditional-fetching": "Conditional Data Fetching",
            "data-fetching": "Data Fetching",
            "error-handling": {
              "display": "hidden",
              "title": "Error Handling",
            },
            "getting-started": "Getting Started",
            "github_link": {
              "href": "https://github.com/shuding/nextra",
              "newWindow": true,
              "title": "GitHub üêô ‚Üí",
              "type": "page",
            },
            "global-configuration": "Global Configuration",
            "middleware": "Middleware",
            "mutation": "Mutation",
            "options": "Options",
            "pagination": "Pagination",
            "prefetching": "Prefetching",
            "raw-layout": {
              "theme": {
                "layout": "raw",
              },
              "title": "Raw Layout",
            },
            "revalidation": "Auto Revalidation",
            "suspense": "Suspense",
            "typescript": "TypeScript :)",
            "with-nextjs": "Next.js SSG and SSR",
            "wrap-toc-items": "Wrap Table of Content Items",
          },
          "kind": "Meta",
          "locale": "en-US",
        },
        {
          "children": [
            {
              "data": {
                "*": {
                  "theme": {
                    "footer": false,
                  },
                },
                "--- yoo": {
                  "title": "Do Not Use",
                  "type": "separator",
                },
                "cache": "Cache",
                "code-highlighting": "Code Highlighting",
                "file-name.with.DOTS": "Filenames with Dots",
                "future-image": "Future Image",
                "markdown-import": "Markdown Import",
                "more": "More: A Super Super Super Super Long Directory",
                "performance": "Performance",
                "react-native": "React Native",
              },
              "kind": "Meta",
              "locale": "en-US",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "cache",
              "route": "/docs/advanced/cache",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "code-highlighting",
              "route": "/docs/advanced/code-highlighting",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "file-name.with.DOTS",
              "route": "/docs/advanced/file-name.with.DOTS",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "future-image",
              "route": "/docs/advanced/future-image",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "markdown-import",
              "route": "/docs/advanced/markdown-import",
            },
            {
              "children": [
                {
                  "kind": "MdxPage",
                  "locale": "en-US",
                  "name": "loooooooooooooooooooong-title",
                  "route": "/docs/advanced/more/loooooooooooooooooooong-title",
                },
                {
                  "children": [
                    {
                      "kind": "MdxPage",
                      "name": "one",
                      "route": "/docs/advanced/more/tree/one",
                    },
                    {
                      "kind": "MdxPage",
                      "name": "three",
                      "route": "/docs/advanced/more/tree/three",
                    },
                    {
                      "kind": "MdxPage",
                      "name": "two",
                      "route": "/docs/advanced/more/tree/two",
                    },
                    {
                      "data": {
                        "one": "One",
                        "three": "Three",
                        "two": "Two",
                      },
                      "kind": "Meta",
                    },
                  ],
                  "kind": "Folder",
                  "name": "tree",
                  "route": "/docs/advanced/more/tree",
                },
                {
                  "data": {
                    "loooooooooooooooooooong-title": "Loooooooooooooooooooong Title",
                    "tree": "Tree",
                  },
                  "kind": "Meta",
                  "locale": "en-US",
                },
              ],
              "kind": "Folder",
              "name": "more",
              "route": "/docs/advanced/more",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "performance",
              "route": "/docs/advanced/performance",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "react-native",
              "route": "/docs/advanced/react-native",
            },
          ],
          "kind": "Folder",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "arguments",
          "route": "/docs/arguments",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "callout",
          "route": "/docs/callout",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "change-log",
          "route": "/docs/change-log",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "code-block-without-language",
          "route": "/docs/code-block-without-language",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "conditional-fetching",
          "route": "/docs/conditional-fetching",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "data-fetching",
          "route": "/docs/data-fetching",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "error-handling",
          "route": "/docs/error-handling",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "getting-started",
          "route": "/docs/getting-started",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "global-configuration",
          "route": "/docs/global-configuration",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "middleware",
          "route": "/docs/middleware",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "mutation",
          "route": "/docs/mutation",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "options",
          "route": "/docs/options",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "pagination",
          "route": "/docs/pagination",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "prefetching",
          "route": "/docs/prefetching",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "raw-layout",
          "route": "/docs/raw-layout",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "revalidation",
          "route": "/docs/revalidation",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "suspense",
          "route": "/docs/suspense",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "typescript",
          "route": "/docs/typescript",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "with-nextjs",
          "route": "/docs/with-nextjs",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "wrap-toc-items",
          "route": "/docs/wrap-toc-items",
        },
      ],
      "kind": "Folder",
      "name": "docs",
      "route": "/docs",
    },
    {
      "children": [
        {
          "data": {
            "auth": "Authentication",
            "basic": "Basic Usage",
            "error-handling": "Error Handling",
            "full": {
              "theme": {
                "layout": "raw",
              },
              "title": "Full",
            },
            "infinite-loading": "Infinite Loading",
            "ssr": "Next.js SSR",
          },
          "kind": "Meta",
          "locale": "en-US",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Authentication",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "auth",
          "route": "/examples/auth",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Basic Usage",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "basic",
          "route": "/examples/basic",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Error Handling",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "error-handling",
          "route": "/examples/error-handling",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "full",
          "route": "/examples/full",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Infinite Loading",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "infinite-loading",
          "route": "/examples/infinite-loading",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Next.js SSR",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "ssr",
          "route": "/examples/ssr",
        },
      ],
      "kind": "Folder",
      "name": "examples",
      "route": "/examples",
    },
    {
      "frontMatter": {
        "searchable": false,
        "title": "React Hooks for Data Fetching",
      },
      "kind": "MdxPage",
      "locale": "en-US",
      "name": "index",
      "route": "/",
    },
    {
      "data": {
        "404": "404",
        "500": "500",
        "about": "About",
        "blog": "Blog",
        "docs": "Docs",
        "examples": "Examples",
      },
      "kind": "Meta",
    },
  ],
  "/docs/data-fetching",
  "Data Fetching",
]
`;

exports[`Page Process > pageMap zh-CN 1`] = `
[
  [
    {
      "data": {
        "about": "About",
        "blog": {
          "title": "ÂçöÂÆ¢",
          "type": "page",
        },
        "docs": {
          "title": "ÊñáÊ°£",
          "type": "page",
        },
        "examples": {
          "title": "Á§∫‰æã",
          "type": "page",
        },
        "index": {
          "display": "hidden",
          "title": "ÁÆÄ‰ªã",
          "type": "page",
        },
      },
      "kind": "Meta",
      "locale": "zh-CN",
    },
    {
      "children": [
        {
          "data": {
            "a-page": {
              "display": "hidden",
              "theme": {
                "layout": "raw",
              },
              "title": "A Page",
              "type": "page",
            },
            "acknowledgement": "üß© Acknowledgement",
            "team": "üë• Team",
          },
          "kind": "Meta",
          "locale": "en-US",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "a-page",
          "route": "/about/a-page",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "acknowledgement",
          "route": "/about/acknowledgement",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "team",
          "route": "/about/team",
        },
      ],
      "kind": "Folder",
      "name": "about",
      "route": "/about",
    },
    {
      "children": [
        {
          "data": {
            "swr-v1": "SWR 1.0 ÂèëÂ∏É",
          },
          "kind": "Meta",
          "locale": "zh-CN",
        },
        {
          "frontMatter": {
            "description": "Â§ßÁ∫¶‰∏§Âπ¥ÂâçÔºåÊàë‰ª¨ÂºÄÊ∫ê‰∫Ü SWRÔºåÂ§ßÂÆ∂ÂñúÁà±ÁöÑÂ∞èÂûãÊï∞ÊçÆËØ∑Ê±Ç React Â∫ì„ÄÇ‰ªäÂ§©ÔºåÊàë‰ª¨ËøéÊù•‰∫ÜÂè¶‰∏Ä‰∏™ÈáåÁ®ãÁ¢ëÔºöSWR 1.0 ÁâàÊú¨ÂèëÂ∏É„ÄÇ",
            "image": "https://assets.vercel.com/image/upload/v1630059453/swr/v1.png",
          },
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "swr-v1",
          "route": "/blog/swr-v1",
        },
      ],
      "kind": "Folder",
      "name": "blog",
      "route": "/blog",
    },
    {
      "children": [
        {
          "data": {
            "advanced": "È´òÁ∫ß",
            "arguments": "‰º†ÂÖ•ÂèÇÊï∞",
            "change-log": "Êõ¥Êñ∞Êó•Âøó",
            "conditional-fetching": "Êù°‰ª∂Êï∞ÊçÆËØ∑Ê±Ç",
            "data-fetching": "Êï∞ÊçÆËØ∑Ê±Ç",
            "error-handling": "ÈîôËØØÂ§ÑÁêÜ",
            "getting-started": "ÂÖ•Èó®",
            "global-configuration": "ÂÖ®Â±ÄÈÖçÁΩÆ",
            "middleware": "‰∏≠Èó¥‰ª∂",
            "mutation": "Êï∞ÊçÆÊõ¥Êîπ",
            "options": "ÈÄâÈ°π",
            "pagination": "ÂàÜÈ°µ",
            "prefetching": "È¢ÑËØ∑Ê±Ç",
            "revalidation": "Ëá™Âä®ÈáçÊñ∞ËØ∑Ê±Ç",
            "suspense": "Suspense",
            "typescript": "Typescript",
            "with-nextjs": "Next.js SSG Âíå SSR",
            "wrap-toc-items": "ÂåÖË£ùÁõÆÈåÑÈ†Ö",
          },
          "kind": "Meta",
          "locale": "zh-CN",
        },
        {
          "children": [
            {
              "data": {
                "cache": "ÁºìÂ≠ò",
                "file-name.with.DOTS": "Â∏¶ÁÇπÁöÑÊñá‰ª∂Âêç",
                "more": "More",
                "performance": "ÊÄßËÉΩ",
                "react-native": "React Native",
              },
              "kind": "Meta",
              "locale": "zh-CN",
            },
            {
              "kind": "MdxPage",
              "locale": "zh-CN",
              "name": "cache",
              "route": "/docs/advanced/cache",
            },
            {
              "kind": "MdxPage",
              "locale": "zh-CN",
              "name": "file-name.with.DOTS",
              "route": "/docs/advanced/file-name.with.DOTS",
            },
            {
              "children": [
                {
                  "children": [
                    {
                      "kind": "MdxPage",
                      "name": "one",
                      "route": "/docs/advanced/more/tree/one",
                    },
                    {
                      "kind": "MdxPage",
                      "name": "three",
                      "route": "/docs/advanced/more/tree/three",
                    },
                    {
                      "kind": "MdxPage",
                      "name": "two",
                      "route": "/docs/advanced/more/tree/two",
                    },
                    {
                      "data": {
                        "one": "One",
                        "three": "Three",
                        "two": "Two",
                      },
                      "kind": "Meta",
                    },
                  ],
                  "kind": "Folder",
                  "name": "tree",
                  "route": "/docs/advanced/more/tree",
                },
                {
                  "kind": "MdxPage",
                  "locale": "en-US",
                  "name": "loooooooooooooooooooong-title",
                  "route": "/docs/advanced/more/loooooooooooooooooooong-title",
                },
                {
                  "data": {
                    "loooooooooooooooooooong-title": "Loooooooooooooooooooong Title",
                    "tree": "Tree",
                  },
                  "kind": "Meta",
                  "locale": "en-US",
                },
              ],
              "kind": "Folder",
              "name": "more",
              "route": "/docs/advanced/more",
            },
            {
              "kind": "MdxPage",
              "locale": "zh-CN",
              "name": "performance",
              "route": "/docs/advanced/performance",
            },
            {
              "kind": "MdxPage",
              "locale": "zh-CN",
              "name": "react-native",
              "route": "/docs/advanced/react-native",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "code-highlighting",
              "route": "/docs/advanced/code-highlighting",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "future-image",
              "route": "/docs/advanced/future-image",
            },
            {
              "kind": "MdxPage",
              "locale": "en-US",
              "name": "markdown-import",
              "route": "/docs/advanced/markdown-import",
            },
          ],
          "kind": "Folder",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "arguments",
          "route": "/docs/arguments",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "change-log",
          "route": "/docs/change-log",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "conditional-fetching",
          "route": "/docs/conditional-fetching",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "data-fetching",
          "route": "/docs/data-fetching",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "error-handling",
          "route": "/docs/error-handling",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "getting-started",
          "route": "/docs/getting-started",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "global-configuration",
          "route": "/docs/global-configuration",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "middleware",
          "route": "/docs/middleware",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "mutation",
          "route": "/docs/mutation",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "options",
          "route": "/docs/options",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "pagination",
          "route": "/docs/pagination",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "prefetching",
          "route": "/docs/prefetching",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "revalidation",
          "route": "/docs/revalidation",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "suspense",
          "route": "/docs/suspense",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "with-nextjs",
          "route": "/docs/with-nextjs",
        },
        {
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "wrap-toc-items",
          "route": "/docs/wrap-toc-items",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "404-500",
          "route": "/docs/404-500",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "advanced",
          "route": "/docs/advanced",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "callout",
          "route": "/docs/callout",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "code-block-without-language",
          "route": "/docs/code-block-without-language",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "raw-layout",
          "route": "/docs/raw-layout",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "typescript",
          "route": "/docs/typescript",
        },
      ],
      "kind": "Folder",
      "name": "docs",
      "route": "/docs",
    },
    {
      "children": [
        {
          "data": {
            "auth": "Ë∫´‰ªΩÈ™åËØÅ",
            "basic": "Âü∫Êú¨Áî®Ê≥ï",
            "error-handling": "ÈîôËØØÂ§ÑÁêÜ",
            "infinite-loading": "Êó†ÈôêÂä†ËΩΩ",
            "ssr": "Next.js SSR",
          },
          "kind": "Meta",
          "locale": "zh-CN",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Ë∫´‰ªΩÈ™åËØÅ",
          },
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "auth",
          "route": "/examples/auth",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Âü∫Êú¨Áî®Ê≥ï",
          },
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "basic",
          "route": "/examples/basic",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "ÈîôËØØÂ§ÑÁêÜ",
          },
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "error-handling",
          "route": "/examples/error-handling",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Êó†ÈôêÂä†ËΩΩ",
          },
          "kind": "MdxPage",
          "locale": "zh-CN",
          "name": "infinite-loading",
          "route": "/examples/infinite-loading",
        },
        {
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "full",
          "route": "/examples/full",
        },
        {
          "frontMatter": {
            "full": true,
            "title": "Next.js SSR",
          },
          "kind": "MdxPage",
          "locale": "en-US",
          "name": "ssr",
          "route": "/examples/ssr",
        },
      ],
      "kind": "Folder",
      "name": "examples",
      "route": "/examples",
    },
    {
      "frontMatter": {
        "title": "Áî®‰∫éÊï∞ÊçÆËØ∑Ê±ÇÁöÑ React Hooks Â∫ì",
      },
      "kind": "MdxPage",
      "locale": "zh-CN",
      "name": "index",
      "route": "/",
    },
    {
      "kind": "MdxPage",
      "name": "404",
      "route": "/404",
    },
    {
      "kind": "MdxPage",
      "name": "500",
      "route": "/500",
    },
    {
      "frontMatter": {
        "searchable": false,
      },
      "kind": "MdxPage",
      "locale": "en-US",
      "name": "blog",
      "route": "/blog",
    },
  ],
  "/docs/data-fetching",
  "Êï∞ÊçÆËØ∑Ê±Ç",
]
`;
